### ISSUE TITLE
Dumping and compiling \special

### ISSUE THREAD

**Issue by aminophen** [aminophen/dviasm#16]
There are problems in dumping/compiling contents of \special.  MWE:

``` tex
\documentclass[a4paper,dvipdfmx]{article}
\usepackage{hyperref}
\usepackage{pxjahyper}

\hypersetup{%
  pdftitle={日本語の表現テスト},
  pdfsubject={カタカナのソ},
  pdfauthor={全角　空白},
  pdfkeywords={キーワードたち},
  pdfdisplaydoctitle=true}

\begin{document}

\section{こことか}
\section{そことか}

\end{document}
```

### pTeX

Japanese characters appearing in \special's are encoded in the internal kanji encoding (Windows: sjis, Unix: eucjp).  However, dviasm option -p (or --ptex) does not know how to decode it.

```
$ platex spcj-p.tex
$ dviasm -p spcj-p.dvi -o spcj-p.dump
Traceback (most recent call last):
  File "/path/to/dviasm", line 1250, in <module>
    aDVI.Load(args[0])
  File "/path/to/dviasm", line 311, in Load
    self.LoadFromFile(fp)
  File "/path/to/dviasm", line 326, in LoadFromFile
    page = self.ProcessPage(fp)
  File "/path/to/dviasm", line 512, in ProcessPage
    q = fp.read(p).decode('utf8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc6 in position 20: invalid continuation byte
```

### upTeX

Japanese characters appearing in \special's are encoded in utf8, and it seems that dviasm can dump it correctly.  However, the compilation into DVI seems corrupt.

```
$ uplatex spcj.tex
$ dviasm spcj.dvi -o spcj.dump
$ dviasm spcj.dump -o spcj.dump.dvi
$ dvipdfmx spcj-p.txt.dvi
spcj.dump.dvi -> spcj.dump.pdf
[1
dvipdfmx:warning: Could not find any valid object.
dvipdfmx:warning: Could not find a value in dictionary object.
dvipdfmx:warning: Dictionary object expected but not found.
dvipdfmx:warning: Interpreting special command docinfo (pdf:) failed.
dvipdfmx:warning: >> at page="1" position="(72, 769.89)" (in PDF)
dvipdfmx:warning: >> xxx "pdf:docinfo<</Title(\xe6\x97\xa5\xe6\x9c\xac\xe8\xaa\x9e\xe3"
dvipdfmx:warning: >> Reading special command stopped around >><</Title(\xe6\x97\xa5\xe6\x9c\xac\xe8\xaa\x9e\xe3\x81\xae\xe.<<
dvipdfmx:fatal: No font selected!

No output PDF file written.
```

**Comment by aminophen** [comment]
It seems that the original Python2 version was OK, since it dumps escape sequences:

```
xxx: 'pdf:docinfo<</Title(\xe6\x97\xa5\xe6\x9c\xac\xe8\xaa\x9e\xe3\x81\xae\xe8\xa1\xa8\xe7\x8f\xbe\xe3\x83\x86\xe3\x82\xb9\xe3\x83\x88)/Subject(\xe3\x82\xab\xe3\x82\xbf\xe3\x82\xab\xe3\x83\x8a\xe3\x81\xae\xe3\x82\xbd)/Creator(LaTeX with hyperref)/Author(\xe5\x85\xa8\xe8\xa7\x92\xe3\x80\x80\xe7\xa9\xba\xe7\x99\xbd)/Keywords(\xe3\x82\xad\xe3\x83\xbc\xe3\x83\xaf\xe3\x83\xbc\xe3\x83\x89\xe3\x81\x9f\xe3\x81\xa1)>>'
```

instead of

```
xxx: 'pdf:docinfo<</Title(日本語の表現テスト)/Subject(カタカナのソ)/Creator(LaTeX with hyperref)/Author(全角\u3000空白)/Keywords(キーワードたち)>>'
```

I guess that contents of \special should not be decoded when DVI -> dump, as we cannot foresee which encoding is used in the input DVI (especially for pTeX DVI).

It seems that the addition of `.decode('utf8')` in 92c814b was inappropriate.

``` python3
      elif o in (XXX1, XXX2, XXX3, XXX4):
        q = fp.read(p).decode('utf8')
        s.append([XXX1, q])
```

However, it is also incorrect to simply remove this `.decode('utf8')`, as the dump output contains a leading "b" that represents a `byte` rather than a `str` object.  Hmm ...

**Comment by aminophen** [comment]
It seems that commit dc55476 fixes the issue (at least for the prepared test cases).

**Comment by aminophen** [comment]
Implemented a new option `-x` (or `--xxx-encoding`) to allow users to specify the correct encoding for dumping and compiling `xxx:` contents (1edde6a).

```
  -x STR, --xxx-encoding=STR
                        encoding for interpreting xxx strings [default=none]
```

Allowed values are: `none` (default), `utf8`, `eucjp`, `sjis`.

* By default, the contents of \special is dumped/compiled as an byte-to-byte escape string, which is the original behavior of Python2 version.
* For pTeX users with ISO-2022-JP-encoded DVI, it would be helpful to set `-x eucjp` (for Unix) or `-x sjis` (for Windows).  For upTeX / XeTeX users, `-x utf8` would be useful.

**Comment by aminophen** [comment]
Released 20200918.
